core/src/main/java/org/apache/iceberg/ReachableFileCleanup.java

```diff
@@ -0,0 +1,152 @@
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.iceberg;
+
+import java.io.IOException;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ExecutorService;
+import java.util.function.Consumer;
+import java.util.stream.Collectors;
+import org.apache.iceberg.exceptions.RuntimeIOException;
+import org.apache.iceberg.io.CloseableIterable;
+import org.apache.iceberg.io.FileIO;
+import org.apache.iceberg.relocated.com.google.common.collect.Sets;
+import org.apache.iceberg.util.Tasks;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * File cleanup strategy for snapshot expiration which determines, via an in-memory reference set,
+ * metadata and data files that are not reachable given the previous and current table states.
+ */
+class ReachableFileCleanup extends FileCleanupStrategy {
+
+  private static final Logger LOG = LoggerFactory.getLogger(ReachableFileCleanup.class);
+
+  ReachableFileCleanup(
+      FileIO fileIO,
+      ExecutorService deleteExecutorService,
+      ExecutorService planExecutorService,
+      Consumer<String> deleteFunc) {
+    super(fileIO, deleteExecutorService, planExecutorService, deleteFunc);
+  }
+
+  @Override
+  public void cleanFiles(TableMetadata beforeExpiration, TableMetadata afterExpiration) {
+    Set<String> manifestListsToDelete = Sets.newHashSet();
+
+    Set<Snapshot> snapshotsBeforeExpiration = Sets.newHashSet(beforeExpiration.snapshots());
+    Set<Snapshot> snapshotsAfterExpiration = Sets.newHashSet(afterExpiration.snapshots());
+    Set<Snapshot> expiredSnapshots = Sets.newHashSet();
+    for (Snapshot snapshot : snapshotsBeforeExpiration) {
+      if (!snapshotsAfterExpiration.contains(snapshot)) {
+        expiredSnapshots.add(snapshot);
+        if (snapshot.manifestListLocation() != null) {
+          manifestListsToDelete.add(snapshot.manifestListLocation());
+        }
+      }
+    }
+
+    Set<ManifestFile> candidateManifestFilesForDeletion = readManifests(expiredSnapshots);
+    Set<ManifestFile> manifestFilesAfterExpiration = readManifests(snapshotsAfterExpiration);
+
+    Set<ManifestFile> manifestsToDelete = Sets.newHashSet(candidateManifestFilesForDeletion);
+    for (ManifestFile candidateManifestFile : candidateManifestFilesForDeletion) {
+      if (manifestFilesAfterExpiration.contains(candidateManifestFile)) {
+        manifestsToDelete.remove(candidateManifestFile);
+      }
+    }
+
+    Set<String> dataFilesToDelete =
+        findFilesToDelete(manifestsToDelete, manifestFilesAfterExpiration);
+    deleteFiles(dataFilesToDelete, "Data");
```

Comment:
Nit: `Data` is capitalized here but no other context strings are capitalized.

Reply:
I've lower cased "data" so it's consistent with the others.